src/handlers/psp.ts

```diff
@@ -0,0 +1,436 @@
+      if (isNil(endpoint)) {
+        node.logger.warn(`EndpointId ${template.endpointId} not found in endpoints`, baseLogOptions);
+        return;
+      }
+      // Verify endpointId
+      const expectedEndpointId = ethers.utils.keccak256(
+        ethers.utils.defaultAbiCoder.encode(['string', 'string'], [endpoint.oisTitle, endpoint.endpointName])
+      );
+      if (expectedEndpointId !== template.endpointId) {
+        node.logger.warn(
+          `EndpointId ${template.endpointId} does not match expected ${expectedEndpointId}`,
+          baseLogOptions
+        );
+        return;
+      }
+
+      groupedSubscriptions.push({
+        subscriptions: enabledSubscriptionsByTemplateId[templateId],
+        template: { ...template, id: templateId },
+        endpoint: { ...endpoint, id: template.endpointId },
+      });
+    });
+  }
+
+  return {
+    config,
+    baseLogOptions,
+    groupedSubscriptions,
+    apiValuesBySubscriptionId: {},
+    providerStates: [],
+  };
+};
+
+const executeApiCalls = async (state: State): Promise<State> => {
+  const { config, baseLogOptions, groupedSubscriptions } = state;
+
+  // TODO: promise.all? 🤔
+  let apiValuesBySubscriptionId: { [subscriptionId: string]: ethers.BigNumber } = {};
+  for (const { subscriptions, template, endpoint } of groupedSubscriptions) {
+    const templateLogOptions: node.LogOptions = {
+      ...baseLogOptions,
+      additional: {
+        templateId: template.id,
+      },
+    };
+    const apiCallParameters = abi.decode(template.templateParameters);
+    const [errorCallApi, logsData] = await retryGo(() =>
+      callApi({
+        oises: config.ois,
+        apiCredentials: config.apiCredentials,
+        apiCallParameters,
+        oisTitle: endpoint.oisTitle,
+        endpointName: endpoint.endpointName,
+      })
+    );
+    if (!isNil(errorCallApi) || isNil(logsData)) {
+      node.logger.warn('Failed to fecth API value', templateLogOptions);
+      continue;
+    }
+    const [logs, apiValue] = logsData;
+    node.logger.logPending(logs, templateLogOptions);
+
+    if (isNil(apiValue)) {
+      node.logger.warn('Failed to fetch API value. Skipping update...', templateLogOptions);
+      continue;
+    }
+
+    apiValuesBySubscriptionId = {
+      ...apiValuesBySubscriptionId,
+      ...subscriptions.reduce((acc, subscription) => ({ ...acc, [subscription.id]: apiValue }), {}),
+    };
+  }
+
+  return { ...state, apiValuesBySubscriptionId };
+};
+
+const initializeProviders = async (state: State): Promise<State> => {
+  const { config, baseLogOptions } = state;
+
+  const airnodeWallet = ethers.Wallet.fromMnemonic(config.nodeSettings.airnodeWalletMnemonic);
+
+  const evmChains = config.chains.filter((chain: ChainConfig) => chain.type === 'evm');
+  if (isEmpty(evmChains)) {
+    throw new Error('One or more evm compatible chain(s) must be defined in the provided config');
+  }
+  const providerPromises = flatMap(
```

Comment:
There is a `flatMap` array function that could replace the lodash `flatMap`. `flatMap` is native so may be more performant. You can also remove the types and replace lodash `map` with the native type: (I think `Object.entries` is a new thing)
```
  const providerPromises =
    evmChains.flatMap(chain =>
      Object.entries(chain.providers).map(async ([providerName, chainProvider]) => {
        const providerLogOptions = {
          ...baseLogOptions,
          meta: {
            ...baseLogOptions.meta,
            chainId: chain.id,
            providerName,
          },
        };
```

Reply:
I was also wondering why we use the lodash functions when there are native functions for this but I didn't want to run into any weird issues so I just did it the same way it's done in [airnode-node](https://github.com/api3dao/airnode/blob/master/packages/airnode-node/src/providers/actions.ts#L53)

Reply:
`airnode-node` is far from perfect :stuck_out_tongue: 
I did some more research and found that lodash functions are 2-3x slower than native but only for _reference_ types, like Object. Types like `number` and `boolean` are faster for lodash weirdly.